Add parameter event filter class

tfoote opened the pull request:
This came out of review of https://github.com/ros2/rclcpp/pull/371#discussion_r146718046

wjwwood commented on rclcpp/include/rclcpp/parameter_client.hpp:
```diff
@@ -268,6 +268,66 @@ class SyncParametersClient
 
 }  // namespace parameter_client
+
+
```
nitpick: minimize vertical whitespace

wjwwood commented on rclcpp/include/rclcpp/parameter_client.hpp:
```diff
@@ -268,6 +268,66 @@ class SyncParametersClient
+   */
+  RCLCPP_PUBLIC
+  inline
+  const std::vector<EventPair> & list()
```
Shouldn't this be a const method too?

wjwwood replied:
Maybe a more descriptive name? Something like `get_events` or `get_filtered_events`.

wjwwood commented on rclcpp/include/rclcpp/parameter_client.hpp:
```diff
@@ -268,6 +268,66 @@ class SyncParametersClient
+  using EventPair = std::pair<EventType, rcl_interfaces::msg::Parameter *>;
+
+  RCLCPP_PUBLIC
+  inline ParameterEventsFilter(
```
I believe the use of `inline` here (and below) is redundant:

> A function defined entirely inside a class/struct/union definition, whether it's a member function or a non-member friend function, is implicitly an inline function.

-- http://en.cppreference.com/w/cpp/language/inline

wjwwood commented on rclcpp/test/test_parameter_events_filter.cpp:
```diff
@@ -0,0 +1,187 @@
+
+  // void TearDown()
+  // {
+  // }
```
remove this?

wjwwood commented on rclcpp/test/test_parameter_events_filter.cpp:
```diff
@@ -0,0 +1,187 @@
+TEST_F(TestParameterEventFilter, full_by_type) {
+  auto res = rclcpp::ParameterEventsFilter(full,
+      {"new", "changed", "deleted"},
+      {nt, ct, dt});
```
nitpick: the indentation here looks like it is one level too many

tfoote replied:
I thought so too, but uncrustify insisted I push it in 4 not 2.

21: --- test/test_parameter_events_filter.cpp
21: +++ test/test_parameter_events_filter.cpp.uncrustify
21: @@ -77,2 +77,2 @@
21: -    {"new", "changed", "deleted"},
21: -    {nt, ct, dt});
21: +      {"new", "changed", "deleted"},
21: +      {nt, ct, dt});
21:

dirk-thomas replied:
The problem is that the code doesn't wrap immediately after the open parenthesis. If you follow that the indentation should be one level only.

tfoote commented:
Rebased after #371 merged and cleaned up implementation and added documenatation.

Linux: [![Build Status](http://ci.ros2.org/job/ci_linux/3560/badge/icon)](http://ci.ros2.org/job/ci_linux/3560/)
Linux aarch64: [![Build Status](http://ci.ros2.org/job/ci_linux-aarch64/758/badge/icon)](http://ci.ros2.org/job/ci_linux-aarch64/758/)
OSX: [![Build Status](http://ci.ros2.org/job/ci_osx/2900/badge/icon)](http://ci.ros2.org/job/ci_osx/2900/)
Windows: [![Build Status](http://ci.ros2.org/job/ci_windows/3645/badge/icon)](http://ci.ros2.org/job/ci_windows/3645/)

tfoote commented:
@ros2/team please review

wjwwood commented on rclcpp/include/rclcpp/parameter_events_filter.hpp:
```diff
@@ -0,0 +1,70 @@
+  /// Construct a filtered view of a parameter event.
+  /**
+   * \param[in] event The parameter event message to filter.
+   * \param[in] names A list of parameter names of interest.
```
This is only a suggestion, but having been away from this code for a while it took me a minute to understand again what the names argument and types arguments were for, so maybe a slightly more in depth explanation of their purpose, perhaps with a simple example, in the body of the doc block would be nice I think.

wjwwood commented on rclcpp/src/rclcpp/time_source.cpp:
```diff
@@ -167,39 +169,28 @@ void TimeSource::clock_cb(const builtin_interfaces::msg::Time::SharedPtr msg)
-          disable_ros_time();
-        }
-      }
+  rclcpp::ParameterEventsFilter filter(event, {"use_sim_time"},
```
This API is a bit unwieldy, which is not a bad thing (at least I can't think of a better way to do it), it's just not obvious at first glance, maybe a comment here to say something like "filter for new or changed events of use_sim_time" would help to make it more obvious when someone is reading this code.

wjwwood commented on rclcpp/src/rclcpp/time_source.cpp:
```diff
@@ -167,39 +169,28 @@ void TimeSource::clock_cb(const builtin_interfaces::msg::Time::SharedPtr msg)
-      parameter_state_ = UNSET;
-    }
+  rclcpp::ParameterEventsFilter deleted(event, {"use_sim_time"},
+    {rclcpp::ParameterEventsFilter::EventType::DELETED});
```
Here it could be "filter everything but the deletion of the use_sim_time parameter".
